waku/v2/waku_relay/protocol.nim

```diff
@@ -131,31 +190,44 @@ iterator subscribedTopics*(w: WakuRelay): lent PubsubTopic =
 proc subscribe*(w: WakuRelay, pubsubTopic: PubsubTopic, handler: WakuRelayHandler) =
   debug "subscribe", pubsubTopic=pubsubTopic
 
-  var subsHandler: PubsubRawHandler
-  when handler is SubscriptionHandler:
-    subsHandler = proc(pubsubTopic: PubsubTopic, data: seq[byte]): Future[void] {.gcsafe.} =
-        let decodeRes = WakuMessage.decode(data)
-        if decodeRes.isErr():
-          debug "message decode failure", pubsubTopic=pubsubTopic, error=decodeRes.error
-          return
+  # rejects messages that are not WakuMessage
+  proc validator(pubsubTopic: string, message: messages.Message): Future[ValidationResult] {.async.} =
```

Comment:
tldr: indeed, done b1df0dd55a42434f45b2fbfe4fad45272574e676

longer version. related to your other comment [here](https://github.com/waku-org/nwaku/pull/1769#pullrequestreview-1455422015) I had the validator inside since I was trying to optimize this, to avoid decoding the message twice (or as many times as validators we have). This optimization was similar to the one in [nimbus](https://github.com/status-im/nimbus-eth2/blob/c22962f3821d069f86ee38826443f9830931cc7c/beacon_chain/networking/eth2_network.nim#L2385-L2386) , where the validator acts also as the handler. Meaning that inside the validator, you also call the handler, and then you dont have to register the handle (just pass nil in subscribe) because the validator calls it for you. Like this you dont have to decode the message i) in the validation stage and ii) in the handler stage, just once.

something like:
```nim
proc validator(pubsubTopic: string, message: messages.Message): Future[ValidationResult] {.async.} =
  let msg = WakuMessage.decode(message.data)
  if msg.isOk():
    yourHandler(yyy) <----
    return ValidationResult.Accept
  return ValidationResult.Reject
```

Buut, at the end I discarded this solution as it has some implications. For example, when having multiple validators (rln, signed topic,e tc). This hacked validator containing also the handler would need to be executed always the last, and this hack on top of a hack seemed too much.

I haven't done a detailed analysis on the impact of decoding the message twice, but haven't seen a significant increase in the simulations on any parameter. So at this stage I would say its overoptimizing. Leaving this idea here in case we want to implement it later on.

Reply:
We have an issue somewhere (https://github.com/status-im/nimbus-eth2/issues/3043) about adding a "user data" field to the Message type, to let the app decode a single time and store the decoded version there.
That will require some trickery since libp2p doesn't know the type of the user data, but feel free to pick that up

Reply:
Ok, very interesting about both possible approaches to avoiding double decoding. Agree with doing what is simple and supported in libp2p for now.